### PR TITLE
Create context menu component

### PR DESCRIPTION
**Pull request by robintown**
This component represents a menu that opens when another UI element is right-clicked or long-pressed.

[Screencast from 2023-10-18 00-56-47.webm](https://github.com/vector-im/compound-web/assets/48614497/f721738a-bbae-4bf5-bf90-e5ee34ab9e29)

[Screencast from 2023-10-18 00-57-04.webm](https://github.com/vector-im/compound-web/assets/48614497/df80f367-3167-428e-9f10-164cddc52682)

### CODE REVIEW

**Review by kerryarchibald — COMMENTED** [review]
Unless I'm missing the key binding, this doesn't seem to be accessible. 
Aside from that, I don't feel super confident this should be added to CW. Right click hijacking is pretty broadly considered a dark pattern. I've raised this with the compound team to see if this is something we want to support.

**Comment by kerryarchibald** [comment]
@robintown as it turns out we already have lots of right click menus in EW Desktop. (I guess right clicking is just not something I do... I had never used them)
They're allowed per platform using `BasePlatform.allowOverridingNativeContextMenus`
Given that right click menus will be disabled in the browser, all their actions *must* be available elsewhere. Which I think means we can get away with these menus being inaccessible. 
https://github.com/matrix-org/design/issues/231

**Comment by t3chguy** [comment]
> Which I think means we can get away with these menus being inaccessible.

Keep in mind right click context menus are also triggered by Accessibility Technologies using the Menu key or Shift + F10, so they should be accessible if they are presented in the accessibility tree whatsoever.

File: src/components/Menu/ContextMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * A menu opened by right-clicking or long-pressing another UI element.
+ */
+export const ContextMenu: FC<Props> = ({
+  title,
+  onOpenChange: onOpenChangeProp,
+  trigger,
+  children: childrenProp,
+}) => {
+  const [open, setOpen] = useState(false);
```
[the remark this answers is not in the document]
Comment (robintown) [reply src/components/Menu/ContextMenu.tsx:72]: Oh, right, I had the same thought! But unfortunately, Radix context menus don't support being controlled - see how [here](https://www.radix-ui.com/primitives/docs/components/context-menu#root) in the docs, the only thing you have access to is an `onOpenChange` prop.

**Comment by robintown** [comment]
Re: accessibility - yes, I think we do need to be careful to only provide this component as an extra convenience for pointing device users, and always provide an alternative route for accessing the menu options. Maybe it's enough to document this in the component's doc comment?

**Comment by t3chguy** [comment]
The current context menus in Element are accessible, given this it is very unlikely we'd want to degrade accessibility by removing those

**Comment by toger5** [comment]
Not sure this is the best place to comment.
In general I am not sure about the blue/black border around the bottom sheet mobile modal. It looks like it has a border at the tob but not left/right. On the current deployment it also has a border on one of the sides and the top.
I thik we should have the border left/right/top or nowhere. Would that make sense for this PR?

**Comment by robintown** [comment]
@t3chguy I've taken a second look at the accessibility concerns:

- The component now sets `aria-haspopup="menu"` on the trigger. Element Web's implementation doesn't do this, but I believe that attribute is applicable here?
- Radix's context menu implementation which we're using here does respond to the Menu key, so as long as the trigger is focusable, we appear to be good on the keyboard navigation front.
- Since the most common use case for context menus is as an additive usability improvement for pointing device users, the extra tab stop often won't be necessary (in the case of Element Web, for instance, it would probably be annoying to have an extra tab stop on every message).
- But still, it's conceivable that a downstream project might end up with a menu that has no other keyboard-accessible entry point, so I've added `hasAccessibleAlternative: boolean` as a required prop. The idea is that this will get consumers to at least acknowledge whether they've taken keyboard accessibility into account, and if not, the component can provide a tab stop as a fallback.

How does this solution sound to you?

**Comment by t3chguy** [comment]
> The component now sets aria-haspopup="menu" on the trigger. Element Web's implementation doesn't do this, but I believe that attribute is applicable here?

Yes, it does. As per MDN:

> The value true is the same as menu

image

Agreed otherwise

**Review by t3chguy — APPROVED** [review]

**Pull request merged**